(D)TLS non-blocking SCR with example

julek-wolfssl opened the pull request:
TLS and DTLS both need to support APP DATA during SCR

julek-wolfssl committed:
DTLS non-blocking scrwith example

julek-wolfssl committed:
TLS and DTLS both need to support APP DATA during SCR

Also some misc fixes

julek-wolfssl committed:
Jenkins fixes

ejohnstown commented:
Is secure renegotiation on the server with a non-blocking socket still disallowed?

julek-wolfssl commented:
I think that if a user wants to enable secure renegotiation, they will have to be prepared to process a APP_DATA_READY error code whether they are using non-blocking sockets or not.
Also, would it make sense to block any non-read API's on a TLS session if the last error code was APP_DATA_READY and the data hasn't been read yet?

ejohnstown commented:
$ ./examples/server/server -u -N -m
    $ ./examples/client/client -u -N -i

The client will do a non-blocking renegotiation just fine, but the server states "not doing secure renegotiation on example with nonblocking yet".

julek-wolfssl committed:
Implement non-blocking SCR on server side

julek-wolfssl commented:
I got the server side implemented. What about:
>would it make sense to block any non-read API's on a TLS session if the last error code was APP_DATA_READY and the data hasn't been read yet?

ejohnstown commented on examples/client/client.c:
```diff
@@ -3089,8 +3090,53 @@ THREAD_RETURN WOLFSSL_THREAD client_test(void* args)
+                    err = wolfSSL_get_error(ssl, 0);
+                    if (err == WOLFSSL_ERROR_WANT_READ ||
+                            err == WOLFSSL_ERROR_WANT_WRITE) {
+                        (void)ClientWrite(ssl,
```
sizeof("...") is including the null at the end of the string.

ejohnstown reviewed (changes requested):
Right now, my opinion is that it is a pending datagram and if not reading it before continuing the handshake means it gets lost, it gets lost.

julek-wolfssl committed:
Don't send null byte

julek-wolfssl committed:
check ClientWrite return

julek-wolfssl commented on src/internal.c:
```diff
@@ -17821,28 +17838,42 @@ int ReceiveData(WOLFSSL* ssl, byte* output, int sz, int peek)
-        WOLFSSL_MSG("Need to start scr, server requested");
-        if ( (err = wolfSSL_Rehandshake(ssl)) != WOLFSSL_SUCCESS)
-            return err;
+        int ret;
```
Accidentally. I restored it.

julek-wolfssl commented on examples/client/client.c:
```diff
@@ -3089,8 +3090,53 @@ THREAD_RETURN WOLFSSL_THREAD client_test(void* args)
+                    if (err == WOLFSSL_ERROR_WANT_READ ||
+                            err == WOLFSSL_ERROR_WANT_WRITE) {
+                        (void)ClientWrite(ssl,
+                                "msg sent during renegotiation",
```
A) Now checking return code
B) Since the RFC requires that processing APP DATA during a renegotiation is supported, this is a way to make sure that we don't lose this support. I can make it an option but the command line options look to be pretty full already.

toddouska replied:
Dong get me wrong, I like the idea of testing App Data during a renegotiation but I would like to avoid ALWAYS ONLY testing App Data during a renegotiation vs the way more common case of not having App Data during.  See if that can easily be added, thanks.

julek-wolfssl replied:
App data during a renegotiation is only tested when using non-blocking sockets. I see that the '-9' command line option. Should I use that or add a value for -i (like './examples/client/client -i test-app-data')?

toddouska replied:
I don't have an opinion on the option name / title.  Always testing APP Data during non-blocking is not good either.  We HAVE to test the most common cases as well as exceptional ones.  Not JUST exceptional ones.

julek-wolfssl replied:
I added an option to `-i` where if the value of this argument is `scr-app-data` then it sends app data. I also patched `mygetopt` so that it will detect if the value being passed to an argument is the next argument.

julek-wolfssl committed:
Passing scr-app-data in to -i to client sends a message during SCR

Modify mygetopt so that if an argument expects a value and that value is the next argument then myoptarg is set to a NULL pointer.
